File: msinv/hull/demography.py
"""Demography for the hull simulator (in generations).

Multi-population demographic models with migration and ms-style
discrete events (population size changes, growth, migration changes,
mergers, splits). Time in generations (NOT coalescent units), to match
the hull simulator's main loop.

Event types
-----------
``('en', t, pop, N)``         Set pop's size to N at time t (going backward).
``('eN', t, N)``              Set ALL pops' sizes to N.
``('eg', t, pop, alpha)``     Set pop's exp-growth rate to alpha.
                              N(t') = N(t) * exp(-alpha * (t' - t))
``('eG', t, alpha)``          Set ALL pops' growth rate to alpha.
``('em', t, dst, src, M)``    Set migration rate from src → dst (per
                              generation, per src lineage) to M.
``('eM', t, M)``              Set ALL off-diagonal migration rates to
                              M / (n_pops - 1).
``('ej', t, src, dst)``       Going backward: at time t every lineage
                              in pop ``src`` moves to pop ``dst``;
                              both populations' migration is zeroed
                              afterward (standard ms convention).

Future work: demestats (Ragsdale 2026) integration as the analytical
rate engine for arbitrary demes graphs.
"""

import math
from typing import Optional


class Demography:
    """Multi-population demographic model in generations.

    Parameters
    ----------
    pop_sizes : list[float]
        Per-population effective sizes at t=0 (present-day).
    migration_matrix : list[list[float]], optional
        ``M[i][j]`` is the per-generation per-src-lineage migration rate
        from population j → i (going backward in time). Diagonals are
        ignored. Defaults to no migration.
    """

    def __init__(
        self,
        pop_sizes: list[float],
        migration_matrix: Optional[list[list[float]]] = None,
    ):
        self.pop_sizes = list(map(float, pop_sizes))
        self.n_pops = len(self.pop_sizes)
        self.growth_rates = [0.0] * self.n_pops
        self.growth_start = [0.0] * self.n_pops
        if migration_matrix is None:
            self.migration_matrix = [[0.0] * self.n_pops for _ in range(self.n_pops)]
        else:
            if len(migration_matrix) != self.n_pops or any(
                len(row) != self.n_pops for row in migration_matrix
            ):
                raise ValueError(
                    f"migration_matrix shape must be {self.n_pops}x{self.n_pops}."
                )
            self.migration_matrix = [list(map(float, row)) for row in migration_matrix]
        # Original events for replaying after a copy() / reset.
        self.events: list[tuple] = []

    def add_event(self, event: tuple):
        """Add a raw ms-style demographic event. Events are kept sorted by time.

        Prefer the named methods below (``add_population_split``,
        ``add_population_size_change``, etc.) for readability.
        """
        self.events.append(event)
        self.events.sort(key=lambda e: e[1])

    def check_connectivity(self, warn: bool = True) -> bool:
        """Verify cross-population coalescence is possible. Returns True
        if connected; False if disjoint components remain after all
        events. When disjoint and ``warn`` is True, emits a warning.

        Lineages in disjoint components never reach a common ancestor
        — downstream Hudson recap via msprime will hang with "infinite
        waiting time until next simulation event".

        Edges considered:
          - migration_matrix nonzero (either direction)
          - 'ej' events (src → dst merge)
          - 'em' events changing migration_matrix at some time
        """
        if self.n_pops <= 1:
            return True
        parent = list(range(self.n_pops))

        def find(x):
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x

        def union(a, b):
            ra, rb = find(a), find(b)
            if ra != rb:
                parent[ra] = rb

        for i in range(self.n_pops):
            for j in range(self.n_pops):
                if i == j:
                    continue
                if self.migration_matrix[i][j] != 0.0:
                    union(i, j)
        for ev in self.events:
            if ev[0] == "ej" and len(ev) >= 4:
                union(int(ev[2]), int(ev[3]))
            elif ev[0] == "em" and len(ev) >= 5 and ev[4] != 0.0:
                union(int(ev[2]), int(ev[3]))
            elif ev[0] == "cmig" and len(ev) >= 7 and ev[6] > 0.0:
                # ('cmig', t, src, dst, kary, inv_id, proportion)
                # Class-conditional migration is also a connectivity edge.
                union(int(ev[2]), int(ev[3]))
            elif ev[0] == "ema" and len(ev) >= 3:
                mat = ev[2]
                for i in range(self.n_pops):
                    for j in range(self.n_pops):
                        if i != j and mat[i][j] != 0.0:
                            union(i, j)
        roots = {find(i) for i in range(self.n_pops)}
        if len(roots) == 1:
            return True
        if warn:
            import warnings as _warnings

            _warnings.warn(
                f"Demography has {len(roots)} disjoint population "
                f"components: {roots}. Lineages across components will "
                f"never reach a common ancestor — msprime recap will "
                f"hang with 'infinite waiting time'. Add migration or "
                f"an 'ej' event.",
                RuntimeWarning,
                stacklevel=2,
            )
        return False

    # -- msprime-compatible convenience methods ----------------------------

    def add_population_split(self, time: float, derived: list[int], ancestral: int):
        """Going backward, merge *derived* populations into *ancestral*.

        Equivalent to msprime ``Demography.add_population_split``
        (and to one ``ej`` event per derived pop in ms).
        """
        for src in derived:
            self.add_event(("ej", time, src, ancestral))

    def add_mass_migration(
        self, time: float, source: int, dest: int, proportion: float = 1.0
    ):
        """Going backward, move *proportion* of lineages from *source*
        to *dest*.

        With proportion=1.0 this is identical to ``ej``.  For fractional
        proportions, use ``add_class_migration`` (or call
        ``add_admixture``) which can also condition on karyotype.
        """
        if proportion != 1.0:
            raise NotImplementedError(
                "Unconditional fractional mass migration (es) not yet "
                "implemented; only the class-conditional version is. "
                "Use add_class_migration(time, src, dst, kary, inv_id, "
                "proportion) or add_admixture(time, src, dst, proportion, "
                "kary=...) — both support proportion < 1."
            )
        self.add_event(("ej", time, source, dest))

    def add_class_migration(
        self,
        time: float,
        source: int,
        dest: int,
        karyotype: str,
        inv_id: int = 0,
        proportion: float = 1.0,
    ):
        """Class-conditional migration / admixture / class-mass-merge.

        Going backward at ``time``, for each lineage currently in
        ``source`` whose karyotype at inversion ``inv_id`` matches
        ``karyotype`` ('S' or 'I'), move it to ``dest`` with probability
        ``proportion``.

        Uses
        ----
        - ``proportion=1.0`` → unconditional class merge.  Models e.g.
          "K's founders at the K-F split were all S karyotype" (forward
          view) by setting ``add_class_migration(t_split, src=K, dst=F,
          karyotype='S')`` — going backward all of K's S lineages join
          F at the split.
        - ``proportion<1.0`` → stochastic admixture / class-conditional
          migration pulse.  Each matching lineage migrates independently
          with the given probability.
        """
        if karyotype not in ("S", "I"):
            raise ValueError(f"karyotype must be 'S' or 'I', got {karyotype!r}")
        if not (0.0 < proportion <= 1.0):
            raise ValueError(f"proportion must be in (0, 1], got {proportion}")
        self.add_event(("cmig", time, source, dest, karyotype, inv_id, proportion))

    def add_admixture(
        self,
        time: float,
        source: int,
        dest: int,
        proportion: float,
        karyotype: Optional[str] = None,
        inv_id: int = 0,
    ):
        """Admixture pulse: at ``time``, fraction ``proportion`` of
        ``source`` migrates into ``dest`` (going backward).

        If ``karyotype`` is given, only lineages of that class at
        ``inv_id`` are eligible.  Otherwise this is currently a synonym
        for ``add_class_migration`` with karyotype='S' (placeholder
        until unconditional fractional migration lands).

        Caveat: with extensive gene flux, some lineages may have all
        of their inv-region segments converted to PAN class.  These
        lineages are NOT caught by cmig (kary check returns None).
        For proportion=1.0 / class-merge use cases, prefer
        ``add_class_split`` which adds a residual ej to catch the
        PAN stragglers.
        """
        if karyotype is None:
            raise NotImplementedError(
                "Class-unconditional admixture (proportion < 1, no "
                "karyotype filter) not yet implemented; supply "
                "karyotype='S' or 'I' to use the class-conditional "
                "operator instead."
            )
        self.add_class_migration(time, source, dest, karyotype, inv_id, proportion)

    def add_class_split(self, time: float, source: int, dest: int, inv_id: int = 0):
        """Class-routed full split: at ``time`` going backward, every
        lineage in ``source`` moves to ``dest``, but the simulator
        routes S- and I-lineages through their class buckets first
        (cmig kary='S', then cmig kary='I'), then mops up any PAN-only
        lineages with a final ej.

        Functionally equivalent to ``add_population_split`` for
        connectivity purposes, but exercises the class-routing code
        path.  Use this if you want to know whether class-routed
        coalescence at the split changes anything compared with
        plain ej (in our K=S-only, F=S+I sampling, it does not).

        Caveats:
        - Final ej is required because gene flux can convert a
          lineage's full inv span to PAN; cmig misses those.
        - With proportion=1.0, the cmig events are unconditional —
          they don't consume RNG (no Bernoulli draw needed).
        """
        self.add_class_migration(time, source, dest, "S", inv_id, 1.0)
        self.add_class_migration(time, source, dest, "I", inv_id, 1.0)
        self.add_event(("ej", float(time), int(source), int(dest)))

    def add_population_size_change(
        self,
        time: float,
        population: Optional[int] = None,
        new_size: Optional[float] = None,
    ):
        """Change effective population size going backward.

        If *population* is None, change all populations (``eN``).
        Otherwise change only that population (``en``).
        """
        if population is None:
            self.add_event(("eN", time, new_size))
        else:
            self.add_event(("en", time, population, new_size))

    def add_growth_rate_change(
        self, time: float, population: Optional[int] = None, growth_rate: float = 0.0
    ):
        """Set exponential growth rate going backward.

        N(t') = N(t) * exp(-growth_rate * (t' - t)).

        If *population* is None, change all populations (``eG``).
        """
        if population is None:
            self.add_event(("eG", time, growth_rate))
        else:
            self.add_event(("eg", time, population, growth_rate))

    def add_migration_rate_change(
        self,
        time: float,
        source: Optional[int] = None,
        dest: Optional[int] = None,
        rate: float = 0.0,
    ):
        """Change migration rate going backward.

        If *source* and *dest* are both None, set all off-diagonal
        migration rates (``eM``).  Otherwise set the single rate
        from *source* to *dest* (``em``; note dest receives migrants
        from source, matching msprime convention).
        """
        if source is None and dest is None:
            self.add_event(("eM", time, rate))
        else:
            self.add_event(("em", time, dest, source, rate))

    def add_inversion_freq_change(
        self, time: float, population: int, inv_id: int, p_inv: float
    ):
        """Change inversion frequency for a specific population at *time*.

        This is useful at population merge events where the ancestral
        population has a different inversion frequency than the derived
        populations (e.g. K has p_inv=0, F has p_inv=0.73, and the
        ancestral pop had p_inv=0.3).

        Parameters
        ----------
        time : float
            Time in generations (going backward).
        population : int
            Population index to modify.
        inv_id : int
            Inversion index (0-based, matching the inversions list order).
        p_inv : float
            New inverted-arrangement frequency for this population.
        """
        self.add_event(("eig", time, population, inv_id, p_inv))

    def copy(self) -> "Demography":
        """Fresh copy with replayable events."""
        d = Demography(list(self.pop_sizes), [list(r) for r in self.migration_matrix])
        d.growth_rates = list(self.growth_rates)
        d.growth_start = list(self.growth_start)
        d.events = list(self.events)
        return d

    # -- size accessors ---------------------------------------------------

    def size_at(self, pop: int, t: float) -> float:
        """Effective size of population ``pop`` at time ``t`` (going
        backward), accounting for current growth rate and the most recent
        size-set event. Does NOT replay events past ``t``.
        """
        if pop >= self.n_pops:
            return 1.0
        N = self.pop_sizes[pop]
        g = self.growth_rates[pop]
        if g == 0.0:
            return N
        return N * math.exp(-g * (t - self.growth_start[pop]))

    # -- event application -------------------------------------------------

    def next_event_time(self, t_now: float) -> float:
        """Time of the next event at or after t_now (or +inf).

        Uses ``>= t_now`` (with a tiny epsilon tolerance) so that
        events scheduled at exactly the current time are still
        returned — important when other event types (class barrier,
        sweep) fire at the same time and advance ``t`` to that value
        without consuming the demographic event.
        """
        for ev in self.events:
            if ev[1] >= t_now - 1e-9:
                return ev[1]
        return float("inf")

    def apply_event_at(self, t: float, active_lineages):
        """Apply all events scheduled for time t in chronological order,
        mutating ``active_lineages`` in place where required.

        Returns a list of (inv_id, pop, p_inv) tuples for any ``eig``
        events that fired — the caller must apply these to its
        InversionSpec list.
        """
        new_events = []
        inv_changes = []
        for ev in self.events:
            if abs(ev[1] - t) > 1e-9:
                new_events.append(ev)
                continue
            etype = ev[0]
            if etype == "eN":
                _, _, N = ev
                for p in range(self.n_pops):
                    # Lock in current size with growth, then reset.
                    self.pop_sizes[p] = N
                    self.growth_rates[p] = 0.0
                    self.growth_start[p] = t
            elif etype == "en":
                _, _, p, N = ev
                if p < self.n_pops:
                    self.pop_sizes[p] = N
                    self.growth_rates[p] = 0.0
                    self.growth_start[p] = t
            elif etype == "eG":
                _, _, alpha = ev
                for p in range(self.n_pops):
                    self.pop_sizes[p] = self.size_at(p, t)
                    self.growth_rates[p] = alpha
                    self.growth_start[p] = t
            elif etype == "eg":
                _, _, p, alpha = ev
                if p < self.n_pops:
                    self.pop_sizes[p] = self.size_at(p, t)
                    self.growth_rates[p] = alpha
                    self.growth_start[p] = t
            elif etype == "eM":
                _, _, M = ev
                if self.n_pops > 1:
                    per = M / (self.n_pops - 1)
                else:
                    per = 0.0
                for i in range(self.n_pops):
                    for j in range(self.n_pops):
                        if i != j:
                            self.migration_matrix[i][j] = per
            elif etype == "em":
                _, _, dst, src, M = ev
                if dst < self.n_pops and src < self.n_pops and dst != src:
                    self.migration_matrix[dst][src] = M
            elif etype == "ej":
                _, _, src, dst = ev
                # Move every active lineage in src to dst.
                for lin in active_lineages:
                    if lin.population == src:
                        lin.population = dst
                # Zero migration to/from src (it no longer exists).
                for k in range(self.n_pops):
                    self.migration_matrix[src][k] = 0.0
                    self.migration_matrix[k][src] = 0.0
            elif etype == "eig":
                _, _, pop, inv_id, p_inv = ev
                inv_changes.append((inv_id, pop, p_inv))
            else:
                raise ValueError(f"Unknown demographic event type: {etype}")
        # Drop processed events.
        self.events = new_events
        return inv_changes

File: rust/msinv-core/src/demography.rs
/// Multi-population demographic model in generations.
///
/// Supports ms-style events: population size changes, exponential
/// growth, migration rate changes, and population mergers.

use crate::lineage::Lineage;

/// Class-conditional migration event "spec" returned alongside
/// inv-changes from `apply_events_at`.  The simulator applies these
/// after the event tick because the handler needs arena access (to
/// read each lineage's class) and an RNG (for fractional proportion).
#[derive(Clone, Debug)]
pub struct ClassMigSpec {
    pub src: u32,
    pub dst: u32,
    pub kary: crate::class_tag::Karyotype,
    pub inv_id: u16,
    pub proportion: f64,
}

/// Demographic event types.
#[derive(Clone, Debug)]
pub enum DemoEvent {
    /// Set ALL pops' sizes to N.
    EN { t: f64, n: f64 },
    /// Set one pop's size to N.
    En { t: f64, pop: u32, n: f64 },
    /// Set ALL pops' exp growth rate.
    EG { t: f64, alpha: f64 },
    /// Set one pop's exp growth rate.
    Eg { t: f64, pop: u32, alpha: f64 },
    /// Set ALL off-diagonal migration rates to M/(n_pops-1).
    EM { t: f64, m: f64 },
    /// Set migration rate from src→dst.
    Em { t: f64, dst: u32, src: u32, m: f64 },
    /// Merge: move all lineages in src into dst (going backward).
    Ej { t: f64, src: u32, dst: u32 },
    /// Change inversion frequency for a specific population.
    Eig { t: f64, pop: u32, inv_id: u16, p_inv: f64 },
    /// Class-conditional migration pulse / mass migration / admixture.
    /// Going backward at time `t`, for each lineage currently in `src`
    /// whose karyotype at `inv_id` matches `kary`, move it to `dst` with
    /// probability `proportion`.
    /// - proportion = 1.0  → unconditional class-mass-merge (like Ej
    ///   but only the matching karyotype moves).  Use to model "K
    ///   founders were S-only at the K-F split" etc.
    /// - proportion < 1.0 → stochastic admixture pulse.  Bernoulli per
    ///   matching lineage.
    /// Requires arena access to read each lineage's class (handled in
    /// the simulator, not in apply_due_events).
    ClassMig { t: f64, src: u32, dst: u32, kary: crate::class_tag::Karyotype,
               inv_id: u16, proportion: f64 },
}

impl DemoEvent {
    pub fn time(&self) -> f64 {
        match self {
            DemoEvent::EN { t, .. } => *t,
            DemoEvent::En { t, .. } => *t,
            DemoEvent::EG { t, .. } => *t,
            DemoEvent::Eg { t, .. } => *t,
            DemoEvent::EM { t, .. } => *t,
            DemoEvent::Em { t, .. } => *t,
            DemoEvent::Ej { t, .. } => *t,
            DemoEvent::Eig { t, .. } => *t,
            DemoEvent::ClassMig { t, .. } => *t,
        }
    }
}

#[derive(Clone, Debug)]
pub struct Demography {
    pub n_pops: u32,
    pub pop_sizes: Vec<f64>,
    pub growth_rates: Vec<f64>,
    pub growth_start: Vec<f64>,
    /// Migration matrix: mig[dst][src] = rate from src → dst.
    pub migration_matrix: Vec<Vec<f64>>,
    pub events: Vec<DemoEvent>,
}

impl Demography {
    pub fn new(pop_sizes: Vec<f64>) -> Self {
        let n = pop_sizes.len() as u32;
        Self {
            n_pops: n,
            pop_sizes: pop_sizes.clone(),
            growth_rates: vec![0.0; n as usize],
            growth_start: vec![0.0; n as usize],
            migration_matrix: vec![vec![0.0; n as usize]; n as usize],
            events: Vec::new(),
        }
    }

    /// Single-population shorthand.
    pub fn single_pop(ne: f64) -> Self {
        Self::new(vec![ne])
    }

    pub fn add_event(&mut self, event: DemoEvent) {
        self.events.push(event);
        self.events.sort_by(|a, b| a.time().partial_cmp(&b.time()).unwrap());
    }

    /// Effective size of `pop` at time `t` (backward), accounting for
    /// current growth rate.
    pub fn size_at(&self, pop: u32, t: f64) -> f64 {
        let p = pop as usize;
        if p >= self.pop_sizes.len() {
            return 1.0;
        }
        let n = self.pop_sizes[p];
        let g = self.growth_rates[p];
        if g == 0.0 {
            return n;
        }
        n * (-g * (t - self.growth_start[p])).exp()
    }

    /// Effective size of `pop` at backward-time `t`, accounting for any
    /// scheduled `En`/`EN`/`Eg`/`EG` events between t=0 (present) and `t`.
    ///
    /// Walks the events list (sorted by ascending time) and applies the
    /// state at the most recent event with time <= t. Required by the
    /// sweep trajectory builder, which iterates forward from `t_origin`
    /// down to `tau` and needs the pop size *as it was at that backward
    /// time*, NOT the current size after subsequent events folded in.
    pub fn pop_size_at(&self, pop: u32, t: f64) -> f64 {
        let p = pop as usize;
        if p >= self.pop_sizes.len() {
            return 1.0;
        }
        let mut size = self.pop_sizes[p];
        let mut growth = self.growth_rates[p];
        let mut growth_start = self.growth_start[p];
        for ev in &self.events {
            if ev.time() > t { break; }
            match ev {
                DemoEvent::EN { n, .. } => { size = *n; growth = 0.0; growth_start = ev.time(); }
                DemoEvent::En { pop: p2, n, .. } if *p2 as usize == p =>
                    { size = *n; growth = 0.0; growth_start = ev.time(); }
                DemoEvent::EG { alpha, .. } => { growth = *alpha; growth_start = ev.time(); }
                DemoEvent::Eg { pop: p2, alpha, .. } if *p2 as usize == p =>
                    { growth = *alpha; growth_start = ev.time(); }
                _ => {}
            }
        }
        if growth == 0.0 { size } else { size * (-growth * (t - growth_start)).exp() }
    }

    /// Time of the next event at or after `t_now`, or +inf.
    pub fn next_event_time(&self, t_now: f64) -> f64 {
        for ev in &self.events {
            if ev.time() >= t_now - 1e-9 {
                return ev.time();
            }
        }
        f64::INFINITY
    }

    /// Apply all events scheduled at time `t`, mutating pop sizes /
    /// growth / migration and moving lineages for merge events.
    /// Returns:
    ///   - inv_changes: (inv_id, pop, p_inv) for any Eig events
    ///   - class_mig: ClassMigSpec list for class-conditional events
    /// The simulator must apply ClassMig specs (needs arena + rng).
    pub fn apply_events_at(&mut self, t: f64, active: &mut [Lineage])
        -> (Vec<(u16, u32, f64)>, Vec<ClassMigSpec>)
    {
        let mut remaining = Vec::new();
        let mut inv_changes: Vec<(u16, u32, f64)> = Vec::new();
        let mut class_mig: Vec<ClassMigSpec> = Vec::new();
        let events = std::mem::take(&mut self.events);
        for ev in events {
            if (ev.time() - t).abs() > 1e-9 {
                remaining.push(ev);
                continue;
            }
            match ev {
                DemoEvent::EN { n, .. } => {
                    for p in 0..self.n_pops as usize {
                        self.pop_sizes[p] = n;
                        self.growth_rates[p] = 0.0;
                        self.growth_start[p] = t;
                    }
                }
                DemoEvent::En { pop, n, .. } => {
                    let p = pop as usize;
                    if p < self.pop_sizes.len() {
                        self.pop_sizes[p] = n;
                        self.growth_rates[p] = 0.0;
                        self.growth_start[p] = t;
                    }
                }
                DemoEvent::EG { alpha, .. } => {
                    for p in 0..self.n_pops as usize {
                        self.pop_sizes[p] = self.size_at(p as u32, t);
                        self.growth_rates[p] = alpha;
                        self.growth_start[p] = t;
                    }
                }
                DemoEvent::Eg { pop, alpha, .. } => {
                    let p = pop as usize;
                    if p < self.pop_sizes.len() {
                        self.pop_sizes[p] = self.size_at(pop, t);
                        self.growth_rates[p] = alpha;
                        self.growth_start[p] = t;
                    }
                }
                DemoEvent::EM { m, .. } => {
                    let per = if self.n_pops > 1 {
                        m / (self.n_pops - 1) as f64
                    } else {
                        0.0
                    };
                    for i in 0..self.n_pops as usize {
                        for j in 0..self.n_pops as usize {
                            if i != j {
                                self.migration_matrix[i][j] = per;
                            }
                        }
                    }
                }
                DemoEvent::Em { dst, src, m, .. } => {
                    let (d, s) = (dst as usize, src as usize);
                    if d < self.n_pops as usize && s < self.n_pops as usize
                        && d != s
                    {
                        self.migration_matrix[d][s] = m;
                    }
                }
                DemoEvent::Ej { src, dst, .. } => {
                    for lin in active.iter_mut() {
                        if lin.population == src {
                            lin.population = dst;
                        }
                    }
                    // Zero migration to/from src.
                    for k in 0..self.n_pops as usize {
                        self.migration_matrix[src as usize][k] = 0.0;
                        self.migration_matrix[k][src as usize] = 0.0;
                    }
                }
                DemoEvent::Eig { pop, inv_id, p_inv, .. } => {
                    inv_changes.push((inv_id, pop, p_inv));
                }
                DemoEvent::ClassMig { src, dst, kary, inv_id, proportion, .. } => {
                    class_mig.push(ClassMigSpec {
                        src, dst, kary, inv_id, proportion
                    });
                }
            }
        }
        self.events = remaining;
        (inv_changes, class_mig)
    }

    /// Compute per-lineage migration rates. Returns (rate, lineage_idx, dst_pop).
    pub fn migration_rates(&self, active: &[Lineage]) -> Vec<(f64, usize, u32)> {
        if self.n_pops < 2 {
            return Vec::new();
        }
        let mut rates = Vec::new();
        for (i, lin) in active.iter().enumerate() {
            let src = lin.population as usize;
            for dst in 0..self.n_pops as usize {
                if dst == src { continue; }
                let m = self.migration_matrix[dst][src];
                if m > 0.0 {
                    rates.push((m, i, dst as u32));
                }
            }
        }
        rates
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn single_pop_size() {
        let d = Demography::single_pop(10_000.0);
        assert_eq!(d.n_pops, 1);
        assert_eq!(d.size_at(0, 0.0), 10_000.0);
        assert_eq!(d.size_at(0, 100.0), 10_000.0); // no growth
    }

    #[test]
    fn exp_growth() {
        let mut d = Demography::single_pop(10_000.0);
        d.growth_rates[0] = 0.001;
        d.growth_start[0] = 0.0;
        // At t=100, N = 10000 * exp(-0.001 * 100) = 10000 * exp(-0.1)
        let expected = 10_000.0 * (-0.1f64).exp();
        assert!((d.size_at(0, 100.0) - expected).abs() < 1.0);
    }

    #[test]
    fn event_ordering() {
        let mut d = Demography::new(vec![1000.0, 1000.0]);
        d.add_event(DemoEvent::Ej { t: 500.0, src: 1, dst: 0 });
        d.add_event(DemoEvent::En { t: 200.0, pop: 0, n: 5000.0 });
        // Should be sorted by time.
        assert_eq!(d.events[0].time(), 200.0);
        assert_eq!(d.events[1].time(), 500.0);
    }

    #[test]
    fn next_event_time() {
        let mut d = Demography::new(vec![1000.0, 1000.0]);
        d.add_event(DemoEvent::Ej { t: 500.0, src: 1, dst: 0 });
        assert_eq!(d.next_event_time(0.0), 500.0);
        assert_eq!(d.next_event_time(600.0), f64::INFINITY);
    }

    #[test]
    fn pop_size_at_walks_en_events() {
        // Forward time: present (t=0) has Ne=1000. At backward t=500, an En
        // event sets Ne=500 (so pre-event size, t > 500, was 500). Going back
        // further past t=500, size should be 500. Below 500, size is 1000.
        let mut d = Demography::new(vec![1_000.0]);
        d.add_event(DemoEvent::En { t: 500.0, pop: 0, n: 500.0 });
        assert!((d.pop_size_at(0, 100.0) - 1_000.0).abs() < 1e-9, "t<event uses current");
        assert!((d.pop_size_at(0, 600.0) - 500.0).abs() < 1e-9, "t>event uses pre-event size");
        assert!((d.pop_size_at(0, 500.0) - 500.0).abs() < 1e-9, "at event uses post-revert size");
    }

    #[test]
    fn pop_size_at_per_pop_independent() {
        let mut d = Demography::new(vec![1_000.0, 2_000.0]);
        d.add_event(DemoEvent::En { t: 500.0, pop: 0, n: 100.0 });
        assert!((d.pop_size_at(0, 600.0) - 100.0).abs() < 1e-9);
        assert!((d.pop_size_at(1, 600.0) - 2_000.0).abs() < 1e-9, "pop 1 untouched");
    }

    #[test]
    fn merge_event_moves_lineages() {
        use crate::segment::{SegmentArena, SEG_NIL};
        let mut d = Demography::new(vec![1000.0, 1000.0]);
        d.add_event(DemoEvent::Ej { t: 100.0, src: 1, dst: 0 });

        let mut arena = SegmentArena::new();
        let seg0 = arena.alloc(0.0, 100.0, 0,
            crate::class_tag::BranchClass::PANMICTIC);
        let seg1 = arena.alloc(0.0, 100.0, 1,
            crate::class_tag::BranchClass::PANMICTIC);
        let mut active = vec![
            Lineage::new(seg0, seg0, 0, 0, &arena),
            Lineage::new(seg1, seg1, 1, 1, &arena),
        ];

        d.apply_events_at(100.0, &mut active);

        // Both lineages should now be in pop 0.
        assert_eq!(active[0].population, 0);
        assert_eq!(active[1].population, 0);
        // Migration from/to pop 1 should be zeroed.
        assert_eq!(d.migration_matrix[0][1], 0.0);
        assert_eq!(d.migration_matrix[1][0], 0.0);
    }
}
